### DISCUSSION TITLE
Schema stitching

### DISCUSSION THREAD

**Discussion by hexwit** [graphql-java/graphql-java#2460] [Q&A]
Hello!

As I know there is no embedded functionality for schema stitching.
It is possible to merge two schemas, but process will fail if types with same name appears in both of them.

I need to merge specific types from both schemas. Merge their field definitions. 
It is Query and Mutation types. Both of them exists in each schema. 

How can I merge types with same names but different fields definitions into one type with 
combined fields? 

I investigated source code, SchemaParser and TypeDefinitionRegistry but didn't find suitable way to 
merge type fields. I have feeling that I am looking at the wrong place.

Could you suggest how to solve my problem, and point me the direction where to dig.

Thanks in advance!

**Comment by bbakerman** [comment]
I think you will want to use SchemaTransformer and a custom visitor that takes the 2nd schema as input and "adds" fields to type as you encounter them, merging them by what ever algo you decide on.

There is currently no doco for this on https://www.graphql-java.com/ but I have a PR up for it here

https://github.com/graphql-java/graphql-java-page/pull/50

and `graphql.schema.SchemaTransformerTest` is another place to show how to do it.

SchemaTransformer takes some grokking (with its `return changeNode` etc..) but its powerful on changing a complex immutable graph that is a schema

**Reply by rajdiran** [reply]
@hexwit Were you able to resolve this issue? I am facing a similar issue now after upgrading the graphql-java version to 15 or more from 11. I have getUser Query defined in multiple schema files and it was working fine before the 15 version. After upgrading it, it is failing with 'Query' extension type [@-1:-1] tried to redefine field 'getUser'

@bbakerman Could you provide me some sample link?

**Reply by bbakerman** [reply]
Can you provide an example of your schema?  defining the same field twice is just not allowed

**Reply by rajdiran** [reply]
@bbakerman  Please find the example schema below,

Module 1

type Query {
    getUser(email: String!): PUserStatus! @auth
}
type PUserStatus{
    status: USER_ONBOARDING_STATUS
}


X Module:

type Query {
	getUser(email: String!) : PUserStatus
}

extend type Query {
    getUser(email: String!) : PUserStatus!
}

type PUserStatus{
    XUserStatus: XUserStatus
}

Y Module:

extend type Query {
    getUser(email: String!) : PUserStatus!
}

type PUserStatus{
    YUserStatus: YUserStatus
}




I have the router project which contains remote schema json for each module. It stitch those schema files and create the GraphqlSchema instance. It was working fine till graphql-java version 14. Now it is failing with 'Query' extension type [@-1:-1] tried to redefine field 'getUser' [@-1:-1]]

**Reply by dondonz** [reply]
That's correct, you are seeing this error message because your Query extension is attempting to redefine a field. This is not permitted in the specification, see this section: https://spec.graphql.org/draft/#sec-Objects.Type-Validation
> For each field of an Object type:
The field must have a unique name within that Object type; no two fields may share the same name.

Perhaps the validation was added in between the versions you are using. An error is the expected behaviour.

Note, version 15 of GraphQL Java is quite out of date. The latest version is 21.x, I'd recommend using a more recent version for security updates.

**Reply by bbakerman** [reply]
Yeah only can defined `type X` and everything else needs to do `extend type X`

I have a beef with the spec in that this does not compose well - eg who gets to be the master definer of the type - any or all?

I would prefer it allowed any number of `extend type X` and they got added together (with field redefinition in place of course)

But this is not how the spec works and hence not how graphql-java now works.

**Reply by rajdiran** [reply]
@dondonz yes I am upgrading the graphql-java to the latest version which 21.x. I just tried different version to identity from which version it stopped working. 

I understand that it is not permitted. is there a way to fix this without updating our existing schema? can custom schema stitch implementation help?

**Reply by dondonz** [reply]
Hi @rajdiran GraphQL Java implements the GraphQL specification, so if it's not permitted in the specification, the library will also not permit it. Perhaps others in the community have more information on schema stitching libraries, however in terms of pure GraphQL Java there isn't a way around it, it's validation to enforce the specification.
